src/PerimeterxFirstPartyClient.php

```diff
@@ -0,0 +1,149 @@
+    const FIRST_PARTY_HEADER_NAME = 'X-PX-First-Party';
+    const FIRST_PARTY_HEADER_VALUE = '1';
+
+    const EMPTY_GIF_B64 = 'R0lGODlhAQABAPAAAAAAAAAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==';
+
+    private $pxConfig;
+    private $logger;
+    private $httpClient;
+
+    /**
+     * @param array $pxConfig
+     * @param Perimeterx\Utils\IHttpClient $httpClient
+     */
+
+    public function __construct($pxConfig, $httpClient) {
+        $this->pxConfig = $pxConfig;
+        $this->logger = $pxConfig['logger'];
+        $this->httpClient = $httpClient;
+    }
+
+    public function handleFirstParty() {
+        $appIdSubstr = substr($this->pxConfig['app_id'], 2);
+        $initJsPath = "/$appIdSubstr/init.js";
+        $xhrPath = "/$appIdSubstr/xhr";
+        $captchaJsPath = "/$appIdSubstr/captcha";
+
+        $uri = strtok($_SERVER['REQUEST_URI'], "?");
+        $method = $_SERVER['REQUEST_METHOD'];
+
+        $retval = null;
+
+        if ($uri === $initJsPath) {
+            $retval = $this->handleFirstPartyClient();
+        } else if (strpos($uri, $xhrPath) !== false) {
+            $retval = $this->handleFirstPartyXHR($xhrPath, $method, $uri);
+        } else if (strpos($uri, $captchaJsPath) !== false) {
+            $retval = $this->handleFirstPartyCaptcha();
+        } else {
+            $this->logger->debug("$method $uri did not match any first party path");
+        }
+        return $retval;
+    }
+
+    private function handleFirstPartyClient() {
+        $this->logger->debug('first party client script detected');
+        return $this->handleFirstPartyScript(self::CLIENT_DOMAIN, "/{$this->pxConfig['app_id']}/main.min.js");
+    }
+
+    private function handleFirstPartyCaptcha() {
+        $this->logger->debug('first party captcha script detected');
+        return $this->handleFirstPartyScript(self::CAPTCHA_DOMAIN, "/{$this->pxConfig['app_id']}/captcha.js");
+    }
+
+    private function handleFirstPartyXHR($xhrPath, $method, $uri) {
+        $this->logger->debug('first party xhr detected');
+
+        $thirdPartyUri = str_replace($xhrPath, '', $uri);
+        $isGifRequest = strpos($thirdPartyUri, ".gif") !== false;
+        $contentType = $isGifRequest ? "image/gif" : "application/json";
+
+        try {
+            $body = $method === 'GET' ? null : PerimeterxUtils::getPostRequestBody();
+            $query = $method === 'GET' ? $_SERVER['QUERY_STRING'] : '';
+            $headers = $this->prepareXhrHeaders();
+            $rawResponse = $this->makeFirstPartyRequest(self::XHR_DOMAIN, $method, $thirdPartyUri, $headers, $query, $body);
+            return $this->returnFirstPartyResponseBasedOn($rawResponse, $contentType);
+        } catch (\Exception $e) {
+            $this->logger->debug("Error handling request to $uri - {$e->getMessage()}");
+            $defaultResponseBody = $isGifRequest ? self::EMPTY_GIF_B64 : '';
+            return $this->returnFirstPartyResponse($contentType, $defaultResponseBody, 200);
```

Comment:
Why not returning 404 or another error code? this request fails this kinda hides the fact that we failed to return the sensor..

Also related to it but not to the PR
Maybe we want to ask for the sensor guys to have a altSesnsorSrc the same as captcha so incase we can't fetch it we will return 404 and they will try to load the sensor from the 3rd party?

Reply:
(1) Always returning 200 is the implementation of the node-core enforcer. I based it on that. (We'll also need more in-depth first party E2E tests and a more thorough explanation in the spec.)
(2) I'm very supportive of an altSensorSrc and actually had the same thought. As you mentioned, this is something out of scope and would need to be added to the snippet.

Reply:
Cool